Add KMS that supports Hashicorp Vault with a ServiceAccount per Tenant

nixpanic opened the pull request:
This new KMS uses a Kubernetes ServiceAccount from a Tenant (Namespace)
to connect to Hashicorp Vault. The provisioner and node-plugin will
check for the configured ServiceAccount and use the token that is
located in one of the linked Secrets. Subsequently the Vault connection
is configured to use the Kubernetes token from the Tenant.

This PR contains the following pieces:

- [x] refactoring of `VaultTokensKMS` to make `vaultTenantConnection` re-usable
- [x] implementation of `VaultTenantSA` KMS
- [x] example `.yaml` files
- [x] e2e tests

Updates: #2222

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

nixpanic committed:
util: VaultTenantSA KMS implementation

This new KMS uses a Kubernetes ServiceAccount from a Tenant (Namespace)
to connect to Hashicorp Vault. The provisioner and node-plugin will
check for the configured ServiceAccount and use the token that is
located in one of the linked Secrets. Subsequently the Vault connection
is configured to use the Kubernetes token from the Tenant.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
util: split vaultTenantConnection from VaultTokensKMS

This makes the Tenant configuration for Hashicorp Vault KMS connections
more modular. Additional KMS implementations that use Hashicorp Vault
with per-Tenant options can re-use the new vaultTenantConnection.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
deploy: allow RBD components to get ServiceAccounts

The provisioner and node-plugin have the capability to connect to
Hashicorp Vault with a ServiceAccount from the Namespace where the PVC
is created. This requires permissions to read the contents of the
ServiceAccount from an other Namespace than where Ceph-CSI is deployed.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
doc: add example for Tenant ServiceAccount

The ServiceAccount "ceph-csi-vault-sa" is expected to be placed in the
Namespace "tenant" so that the provisioner and node-plugin fetch the
ServiceAccount from a Namespace where Ceph-CSI is not deployed.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
e2e: add test for Vault with ServiceAccount per Tenant

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
e2e: add securityContext.runAsUser to vault-init-job

Kubelet sometimes reports the following error:

    failed to "StartContainer" for "vault-init-job" with CreateContainerConfigError: container has runAsNonRoot and image will run as root

Setting securityContext.runAsUser resolves this.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

humblec commented on deploy/rbd/kubernetes/csi-provisioner-rbac.yaml:
```diff
@@ -55,6 +55,9 @@ rules:
+    resources: ["serviceaccounts"]
+    verbs: ["get"]
 ---
 kind: ClusterRoleBinding
```
@nixpanic do we need rbac for both provisioner and nodeplugin ? :thinking:

nixpanic replied:
yes, both need to connect to the KMS

humblec replied:
the design doc covers the `createVolume` flow (ie provisioner) part here https://github.com/ceph/ceph-csi/pull/2227/files#diff-43620beda90c9977d037956da8ade411e16b43a29efed8b37ca46c0b78834b61R79 , but the node part has not covered over there, do you think we can enhance and document `plugin` part as well in the doc ?

humblec commented on internal/util/vault_sa.go:
```diff
@@ -0,0 +1,246 @@
+	kms.ConfigName = vaultTokensDefaultConfigName
+	kms.tenantSAName = vaultTenantSAName
+
+	// TODO: should this be configurable per tenant?
```
I think if the vault namespaces are in place, this looks to be a general configuration, will confirm it though

nixpanic replied:
@humblec this was still waiting for a response from you. I have decided that tenants should be able to configure the `vaultRole` option too, it may be needed for tenants to override the default or global configured option.

humblec commented on internal/util/vault_sa.go:
```diff
@@ -0,0 +1,246 @@
+// configuration if it has been set in the options.
+func (kms *VaultTenantSA) setServiceAccountName(config map[string]interface{}) error {
+	err := setConfigString(&kms.tenantSAName, config, "tenantSAName")
+	if errors.Is(err, errConfigOptionInvalid) {
```
can we also log this error ?

nixpanic replied:
it is logged, the calling functions will take care of that

Madhu-1 commented on charts/ceph-csi-rbd/templates/nodeplugin-clusterrole.yaml:
```diff
@@ -22,4 +22,7 @@ rules:
   - apiGroups: [""]
     resources: ["configmaps"]
     verbs: ["get"]
+  - apiGroups: [""]
```
This should go to nodeplugin role instead of clusterrole?

nixpanic replied:
it needs to be a ClusterRole, the ServiceAccount is located in the Tenants namespace.

Madhu-1 commented on e2e/deploy-vault.go:
```diff
@@ -91,3 +97,79 @@ func createORDeleteVault(action string) {
+
+// waitForJobCompletion polls the status of the given job and waits until the
+// jobs has succeeded or until the timeout is hit.
+func waitForJobCompletion(c kubernetes.Interface, ns, job string, timeout int) error {
```
move this to utils.go or add a new file for job-specific helper functions.

nixpanic replied:
moved to utils.go

Madhu-1 commented on e2e/deploy-vault.go:
```diff
@@ -91,3 +97,79 @@ func createORDeleteVault(action string) {
+	const jobName = "vault-tenant-sa"
+	err := waitForJobCompletion(c, cephCSINamespace, jobName, deployTimeout)
+	if err != nil {
+		e2elog.Failf("Job %s/%s did not succeed: %v", cephCSINamespace, jobName, err)
```
return an error from here and let the caller handle it?

nixpanic replied:
done!

Madhu-1 commented on internal/util/vault_sa.go:
```diff
@@ -0,0 +1,246 @@
+	c := kms.getK8sClient()
+	sa, err := c.CoreV1().ServiceAccounts(kms.Tenant).Get(context.TODO(),
+		kms.tenantSAName, metav1.GetOptions{})
+
```
newline not required?

nixpanic replied:
removed.

Madhu-1 commented on internal/util/vault_sa.go:
```diff
@@ -0,0 +1,246 @@
+		return "", err
+	}
+
+	err = ioutil.WriteFile(dir+"/token", []byte(token), 0600)
```
define a new variable for `dir+/token`? as it is used in multiple places?

nixpanic replied:
Only used 2x, in the same function. I do not think it is needed or making things clearer when defining a constant for it.

Madhu-1 replied:
yes currently used only in 2 places. if something changes in the future same changes need to be made in multiple places

Madhu-1 reviewed (commented):
small nit. LGTM

Madhu-1 commented:
@nixpanic E2E is failing can you please take a look at it.

nixpanic commented:
> @nixpanic E2E is failing can you please take a look at it.

Yes, the new `tenants-sa-admin.yaml` configuration seems to break the existing Vault config. I'm trying to understand how a single Vault service can work with different ServiceAccounts, roles and namespaces, and how to apply those while running the e2e suite.

Thanks for the reviews!

humblec commented:
LGTM.. will just for the `e2e` to pass and approve.

nixpanic committed:
util: make parseTenantConfig() usable for modular KMSs

parseTenantConfig() only allowed configuring a defined set of options,
and KMSs were not able to re-use the implementation. Now, the function
parses the ConfigMap from the Tenants Namespace and returns a map with
options that the KMS supports.

The map that parseTenantConfig() returns can be inspected by the KMS,
and applied to the vaultTenantConnection type by calling parseConfig().

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic commented:
Commit `util: make parseTenantConfig() usable for modular KMSs` has been added to allow the Tenants ConfigMap to contain a different `vaultRole` than the default or global configuration. This also might help :crossed_fingers: to configure multiple ServiceAccounts in Vault.

Madhu-1 commented:
/retest ci/centos/mini-e2e/k8s-1.21

Madhu-1 commented:
> /retest ci/centos/mini-e2e/k8s-1.21

`X Exiting due to GUEST_PROVISION: Failed to start host: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute
`

nixpanic commented:
/retest ci/centos/mini-e2e/k8s-1.21

nixpanic commented:
> /retest ci/centos/mini-e2e/k8s-1.21

etcdserver timeout, [logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.21/detail/mini-e2e_k8s-1.21/259/pipeline):
```
Jul 13 14:38:02.316: INFO: waiting for PVC rbd-pvc (18 seconds elapsed)
Jul 13 14:38:16.008: INFO: Error getting pvc in namespace: '': etcdserver: request timed out
Jul 13 14:38:16.008: FAIL: failed to create PVC with error failed to get pvc: etcdserver: request timed out
```
